File: docs/test-cases/broken-access-control.java
// Test case: broken-access-control (A01:2025)
// Spring Boot controller with IDOR, missing role gate, and SSRF.
package com.example.shop;

import java.net.URL;
import java.net.HttpURLConnection;
import java.io.InputStream;
import org.springframework.web.bind.annotation.*;
import org.springframework.beans.factory.annotation.Autowired;

@RestController
public class OrderController {

    @Autowired private OrderRepository orders;
    @Autowired private UserRepository users;

    // BUG: IDOR — looks up order by path id with no ownership check.
    // Any authenticated user can read any other user's order.
    @GetMapping("/orders/{id}")
    public Order getOrder(@PathVariable Long id) {
        return orders.findById(id).orElse(null);
    }

    // BUG: admin endpoint missing @PreAuthorize("hasRole('ADMIN')").
    // Vertical privilege escalation — any authenticated caller can delete users.
    @DeleteMapping("/admin/users/{userId}")
    public void deleteUser(@PathVariable Long userId) {
        users.deleteById(userId);
    }

    // BUG: SSRF — fetches arbitrary user-supplied URL with no allowlist.
    // Attacker can pivot to internal services (169.254.169.254, localhost, etc.).
    @GetMapping("/preview")
    public String preview(@RequestParam("url") String userUrl) throws Exception {
        URL u = new URL(userUrl);
        HttpURLConnection conn = (HttpURLConnection) u.openConnection();
        try (InputStream in = conn.getInputStream()) {
            return new String(in.readAllBytes());
        }
    }

    // BUG: ownership check happens AFTER the destructive action.
    @PostMapping("/orders/{id}/cancel")
    public void cancel(@PathVariable Long id, @RequestHeader("X-User-Id") Long uid) {
        orders.markCancelled(id);
        Order o = orders.findById(id).orElse(null);
        if (o != null && !o.getOwnerId().equals(uid)) {
            throw new RuntimeException("not yours");
        }
    }
}

File: docs/test-cases/broken-access-control.go
// Test case: broken-access-control (A01:2025)
// net/http handlers with IDOR, env-flag admin gate, and SSRF.
package main

import (
	"encoding/json"
	"io"
	"net/http"
	"os"
	"strings"
)

func getInvoice(w http.ResponseWriter, r *http.Request) {
	id := strings.TrimPrefix(r.URL.Path, "/invoices/")
	// BUG: IDOR — fetches invoice by id with no check that the
	// authenticated caller actually owns it.
	inv, _ := db.QueryRow("SELECT * FROM invoices WHERE id = ?", id)
	json.NewEncoder(w).Encode(inv)
}

func adminPurge(w http.ResponseWriter, r *http.Request) {
	// BUG: admin route gated only by an env flag toggle, not by the
	// caller's role. Anyone who can reach the route on a server with
	// ADMIN_ENABLED=1 can purge data.
	if os.Getenv("ADMIN_ENABLED") != "1" {
		http.Error(w, "disabled", 403)
		return
	}
	db.Exec("DELETE FROM users")
	w.Write([]byte("ok"))
}

func fetchPreview(w http.ResponseWriter, r *http.Request) {
	userURL := r.URL.Query().Get("url")
	// BUG: SSRF — http.Get on caller-supplied URL with no allowlist.
	resp, err := http.Get(userURL)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	w.Write(body)
}

func main() {
	http.HandleFunc("/invoices/", getInvoice)
	http.HandleFunc("/admin/purge", adminPurge)
	http.HandleFunc("/preview", fetchPreview)
	http.ListenAndServe(":8080", nil)
}
